### PR TITLE
add example "do_checkout_ref"

### PR DESCRIPTION
**Pull request by antis81**
@ethomson Here it is! My first example contribution. Not yet compiling! Needs some discussion!

### CODE REVIEW

**Comment by antis81** [comment]
Answered the questions by myself -> deleted them as they are not relevant. Yay, I'm learning C :smile:!

The two `common.h` files are clashing - one in `examples` and one in `examples/network`. The latter should be renamed, maybe `remote-funcs.h` or simply `examples-network.h`. And ... maybe they should be put together in a `common` folder. Why? You might have noticed the copied `PRIuz` macro definition. This has to be cleaned up!

Appart from that, examples are not compiling because of a missing "main()" function. Any hint on that?

**Comment by ashisdhimal** [comment]
haahaahha

**Comment by antis81** [comment]
@ashisdhimal Funny. :smile: 

Anyways. Anyone has a hint on why the compiler moans about the missing main()?

**Comment by Therzok** [comment]
I don't want to seem lazy, but here's a [SO](http://stackoverflow.com/questions/14919366/how-to-compile-library-on-c-using-gcc) link.

**Comment by Therzok** [comment]
You need to integrate your files with the build process in here and use `make`.

See [Makefile](https://github.com/libgit2/libgit2/blob/master/examples/Makefile) here.

**Comment by Therzok** [comment]
Another example of integration in https://github.com/libgit2/libgit2/commit/56ec2256f24eb43b09351194d02cd8458ff39708.

**Comment by antis81** [comment]
Thanks. Also this is better to add to CMake isn't it?

**Comment by Therzok** [comment]
@antis81 CMake already globs all the .c files.

**Comment by antis81** [comment]
>@antis81 CMake already globs all the .c files.

Yep, but it does not create the targets. We should define them to CMake as well instead of writing additional `Makefiles`. Shouldn't we remove the additional Makefiles completely? A short "HowTo add an example" would be nice.

**Comment by antis81** [comment]
@ethomson Long time. Would be nice hearing from you - helping to get this one finished.

File: examples/checkout.c [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+# include <pthread.h>
+# include <unistd.h>
+#endif
+
+
+typedef struct progress_data {
+    size_t completed_steps;
+    size_t total_steps;
+    const char *path;
+} progress_data;
```
Comment (pks-t) [inline examples/checkout.c]: In general, we indent code by tabs and not spaces

File: examples/checkout.c [hunk 2]
```diff
@@ -0,0 +1,102 @@
+#include "common.h"
+#include <git2.h>
+#include <git2/checkout.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#ifndef _WIN32
+# include <pthread.h>
+# include <unistd.h>
+#endif
```
Comment (pks-t) [inline examples/checkout.c]: Are these platform-specific headers even required? I can't see anything making use of them.

File: examples/checkout.c [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+static void print_progress(const progress_data *pd)
+{
+    int checkout_percent = pd->total_steps
+                           ? (100 * pd->completed_steps / pd->total_steps)
+                           : 100;
+
+    printf( "chk %3d%% (%4" PRIuZ "/%4" PRIuZ ") %s\n",
+            checkout_percent,
+            pd->completed_steps, pd->total_steps,
+            pd->path );
```
Comment (pks-t) [inline examples/checkout.c]: I feel like this function could easily be moved into its caller. Like this, you could even get rid of the payload completely, as all infos you need are already being provided to the callback

File: examples/checkout.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+int do_checkout_ref(git_repository *repo, git_reference *checkout_ref)
+{
+    progress_data pd = {0};
+    git_object *target_tree = NULL;
+    git_checkout_options checkout_opts = GIT_CHECKOUT_OPTIONS_INIT;
+
+    int error;
+
+    // Set up options
```
Comment (pks-t) [inline examples/checkout.c]: libgit2 is written in strict C90, so C++-style comments are not allowed

File: examples/checkout.c [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    checkout_opts.checkout_strategy = GIT_CHECKOUT_SAFE | GIT_CHECKOUT_ALLOW_CONFLICTS;
+    checkout_opts.progress_cb = checkout_progress;
+    checkout_opts.progress_payload = &pd;
+
+    // Get the tree for checkout
+    error = git_reference_peel( &target_tree, checkout_ref, GIT_OBJ_TREE );
+    if ( error ) {
+        goto error;
+    }
```
Comment (pks-t) [inline examples/checkout.c]: Please avoid spaces in braces. So it should rather be `if (error)`. Furthermore, there is no need for curly braces around single-line statements.

File: examples/checkout.c [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Do the checkout
+    error = git_checkout_tree( repo, target_tree, &checkout_opts );
+    if ( error ) {
+        goto error;
+    }
+
+    // Update HEAD to point to the reference
```
Comment (pks-t) [inline examples/checkout.c]: See above, this comment doesn't really help much

File: examples/checkout.c [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+int main(int argc, char **argv)
+{
+        git_repository *repo = NULL;
+
+        if (argc != 1 || argv[1] /* silence -Wunused-parameter */)
+                fatal("Sorry, no options supported yet", NULL);
+
+        // TODO: lookup a provided reference name in the repo
+        git_reference  *ref = NULL; // provide a valid reference
```
Comment (pks-t) [inline examples/checkout.c]: What you're searching for is `git_reference_dwim`. ;) Also, you need to avoid variable declarations after statements.

**Comment by tiennou** [comment]
Closing, as it seems to have been superseded by #4692. Thanks for the initial PR !

**Pull request closed without merging**